File: Java/MET-TP1/src/hierarchie/ZeroRec.java
package hierarchie;

public class ZeroRec extends Zero {
	public final static FabriqueNaturels<Nat> FAB = new ZeroRec();
	
	public ZeroRec() {
		
	}
	
	public Nat somme(Nat x) {
		if (x.estNul()) {
			return this;
		} else {
			return this.creerSuccesseur(this.somme(x.predecesseur()));
		}
	}
	

	@Override
	public Nat creerNatAvecValeur(int val) throws IllegalArgumentException {
		if (this.val() < 0) {
			throw new UnsupportedOperationException();
		}
		return SuccRec.FAB.creerNatAvecValeur(val);
	}

	@Override
	public Nat creerZero() {
		return new ZeroRec();
	}

	@Override
	public Nat creerSuccesseur(Nat predecesseur) {
		return new SuccRec(predecesseur);
	}

	@Override
	public Nat creerNatAvecRepresentation(String repDecimale) throws IllegalArgumentException {
		if (!(repDecimale.matches("\\d+"))) {
			throw new IllegalArgumentException();
		} else if (repDecimale.length() > 1 && repDecimale.charAt(0) == '0') {
			throw new IllegalArgumentException();
		}
		return SuccRec.FAB.creerNatAvecRepresentation(repDecimale);
	}
	
}

File: TypeScript/hierarchie/ZeroRec.ts
import { Nat } from "./Nat";
import { NatParInt } from "./NatParInt";
import { FabriqueNaturels } from "./FabriqueNaturels";
import { Zero } from "./Zero";
import { SuccRec } from "./Succrec";

export class ZeroRec extends Zero{
    public static FAB:FabriqueNaturels<Nat> = new ZeroRec(0);

    constructor(a: number) {
        super(0);
    }

    somme(x:Nat):Nat{
		console.log("somme Z " + this.toString() + " " + x.toString());
        if (x.estNul()){
            return this;
        } else {
            return this.creerSuccesseur(this.somme(x.predecesseur()));
        }
        
    }

    creerNatAvecValeur(val:number):Nat{
        return SuccRec.FAB.creerNatAvecValeur(val);
    }
    creerZero():Nat{
        return new ZeroRec(0);
    }
    creerSuccesseur(predecesseur:Nat):Nat{
        return SuccRec.FAB.creerSuccesseur(predecesseur);
    }
    creerNatAvecRepresentation(repDecimale):Nat{
        return SuccRec.FAB.creerNatAvecRepresentation(repDecimale);
    }







}